### DISCUSSION TITLE
Seq Apps & package references - which to include?

### DISCUSSION THREAD

**Discussion by jcmrva** [datalust/seq-tickets#1904] [Help] [answered]
I've created a simple app, targeting `netstandard2.0`, that initially had these 2 references:
```
    <PackageReference Include="Seq.Apps" Version="2021.4.0" />
    <PackageReference Include="System.Net.Http.Json" Version="7.0.1" />
```
This installs and runs fine on my local Seq (2023.3) but fails and runtime on our server (2022.1) with:
> System.IO.FileNotFoundException: Could not load file or assembly 'System.Text.Json, Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. The system cannot find the file specified.

I tried adding this but it did not make a difference:
```
    <PackageReference Include="System.Text.Json" Version="7.0.3" />
```
I have no prior experience building Nuget packages and finding a general answer has been confusing.

How can other Nuget packages (or dlls directly) be included so that Seq can load them?
Is there a methodical way of determining which will be needed - without trial and error or duplicating the production environment locally?

**Comment by nblumhardt** [comment, marked as answer]
Hi Josh, thanks for dropping us a line.

The _Seq.App.HttpRequest_ project is the best example to follow for this; it has a .CSPROJ file with added `<Include` element that pulls in binaries for referenced dependencies:

https://github.com/datalust/seq-app-httprequest/blob/dev/src/Seq.App.HttpRequest/Seq.App.HttpRequest.csproj#L31

The Build.ps1 file in that repository then shows how the package is built.

Seq App packages are a little different from other NuGet packages because of the additional restriction that they don't download dependencies at install time, so this extra bit of packaging logic is required to take care of that.

Let me know if this sorts it out for you,
Nick

**Reply by jcmrva** [reply]
I was able to get the dlls included in the package, but now I get a different exception when running it.
> System.IO.FileLoadException: Could not load file or assembly 'System.Text.Json, Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. Could not find or load a specific file. (0x80131621)
File name: 'System.Text.Json, Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'
 ---> System.IO.FileLoadException: Could not load file or assembly 'System.Text.Json, Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at System.Runtime.Loader.AssemblyLoadContext.InvokeResolveEvent(ResolveEventHandler eventHandler, RuntimeAssembly assembly, String name)
   at System.Runtime.Loader.AssemblyLoadContext.OnAssemblyResolve(RuntimeAssembly assembly, String assemblyFullName)
   at (my code)

`System.Reflection.Assembly.GetExecutingAssembly().Location` returns `/data/Packages/nugetfeed-1056/MyApp/1.0.0/lib/netstandard2.0/MyApp.dll` and I can see all the included dlls in that location.

Our production Seq server is the container version, if that helps.

Is it possible I need to downgrade the package version?

**Reply by nblumhardt** [reply]
Great! This is possibly because the server version of Seq that you're running (2022.1) is based on .NET 6, but your app is (somehow) using .NET 7 features, which Seq 2023.x supports.

Upgrading the server will be your best bet, if that's possible?

**Reply by jcmrva** [reply]
I've requested it. In the meantime, I'll try an earlier version of System.Text.Json - although I would've expected any (supposedly!) netstandard2.0 library to work.

**Reply by jcmrva** [reply]
The version downgrade did work, once I pinned *everything*.
